flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/StrToDateCallGen.scala

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.codegen.calls
+
+import org.apache.flink.api.common.typeinfo.SqlTimeTypeInfo
+import org.apache.flink.table.codegen.calls.CallGenerator.generateCallIfArgsNotNull
+import org.apache.flink.table.codegen.{CodeGenerator, GeneratedExpression}
+import org.apache.flink.table.runtime.functions.DateTimeFunctions
+
+/**
+  * Generates StrToDate call based on the type of last operand.
+  */
+class StrToDateCallGen extends CallGenerator {
+  override def generate(codeGenerator: CodeGenerator,
+                        operands: Seq[GeneratedExpression]): GeneratedExpression = {
+    if (operands.last.literal) {
```

Comment:
@twalthr It seems that we cannot get the value of literals from GeneratedExpression, what do you mean about the offer of calcite ?

Reply:
Yes, we cannot get the value of literals "yet". We can change this in our code to make it work. But the question is how would this work for Calcite's type system?

Reply:
@twalthr What do you think of adding an attribute "literalValue" in GeneratedExpression? I'm not sure if calcite's type system needs this, because GeneratedExpression is a flink's class.

Reply:
@buptljy Yes, I meant something like a `literalValue` for `GeneratedExpression` or sth. more object-oriented. A `LiteralExpression` that extends `GeneratedExpression`. However, before we do this. We need to make sure that Calcite supports this function. We cannot return a different type as Calcite. We need to stay in sync with Calcite's type system otherwise we get a type mismatch in following operators. So first we should explore if Calcite supports this functions with its operand type/return type inference logic.

Reply:
@twalthr Could you show an example of the conflicts between LiteralExpression and calcite type system? From my perspective, the GeneratedExpression is used in the generation of flink functions like what we do in ExpressionReducer: line 95. If the current GeneratedExpression is okay, why won't the child class LiteralExpression be okay?
Sorry I still can't get your points.

Reply:
In the CodeGenerator line 744 you can see that we calculate the result type from Calcite:
```
val resultType = FlinkTypeFactory.toTypeInfo(call.getType)
```

We cannot simply return our own type for functions, we always need to stay in sync with Calcite. For following operators such as:

```
1. DataStreamCalc(field: DATE) # if we would return TIMESTAMP here, the following operation would still assume DATE, thus we get a type mismatch.
2. DataStreamCalc((field + 1): DATE)
```

Reply:
@twalthr Thanks for pointing it out. But we don't have the problem here, because I wrote ScalarSqlFunctionReturnTypes to return different types according to operands types, which is supported by Calcite.

Reply:
But does your return logic looks at the actual value of the operand, no. And this would be required if we want to have same functionality as MySQL's function.

Reply:
Sorry, I didn't see you PR update. Will take another look soon :)